File: structural/composite/composite.cs
using System;
using System.Collections.Generic;

class CompositePattern
{
    abstract class AnimalGangComponent
    {
        public virtual void Add(AnimalGangComponent component) { }

        public virtual void GetDescription() { }
    }

    class Animal : AnimalGangComponent
    {
        private string name { get; }
        private string species { get; }
        private string breed { get; }

        public Animal(string name, string species, string breed)
        {
            this.name = name;
            this.species = species;
            this.breed = breed;
        }

        public override void GetDescription()
        {
            Console.WriteLine("{0} {1} {2}.", breed, species, name);
        }
    }

    class AnimalGang : AnimalGangComponent
    {
        private string name { get; }
        private List<AnimalGangComponent> animals { get; } = new List<AnimalGangComponent>();

        public AnimalGang(string name)
        {
            this.name = name;
        }

        public override void Add(AnimalGangComponent component)
        {
            animals.Add(component);
        }

        public override void GetDescription()
        {
            Console.WriteLine("Animal gang {0}.", name);

            animals.ForEach(animal => {
                animal.GetDescription();
            });
        }
    }

    static void Main(string[] args)
    {
        Animal simba = new Animal("Simba", "cat", "Siamese");
        Animal oscar = new Animal("Oscar", "cat", "Maine Coon");
        Animal oliver = new Animal("Oliver", "cat", "British Shorthair");
        Animal elvis = new Animal("Elvis", "dog", "Siberian Husky");
        Animal buddy = new Animal("Buddy", "dog", "Retriever");
        Animal bruno = new Animal("Bruno", "dog", "Alabai");

        AnimalGang mainGang = new AnimalGang("Animals");
        AnimalGang catGang = new AnimalGang("Cats");
        AnimalGang dogGang = new AnimalGang("Dogs");

        mainGang.Add(simba);
        mainGang.Add(elvis);
        catGang.Add(oscar);
        catGang.Add(oliver);
        mainGang.Add(catGang);
        dogGang.Add(buddy);
        dogGang.Add(bruno);
        mainGang.Add(dogGang);

        mainGang.GetDescription();
    }
}

File: structural/composite/composite.js
class AnimalGangComponent {
    add(component) { }

    getDescription() { }
}

class Animal extends AnimalGangComponent {
    constructor(name, species, breed) {
        super();
        this._name = name;
        this._species = species;
        this._breed = breed;
    }

    getDescription() {
        console.log(`${this._breed} ${this._species} ${this._name}.`);
    }
}

class AnimalGang extends AnimalGangComponent {
    constructor(name) {
        super();
        this._name = name;
        this._animals = [];
    }

    add(component) {
        this._animals.push(component);
    }

    getDescription() {
        console.log(`Animal gang "${this._name}".`);

        this._animals.forEach(animal => {
            animal.getDescription();
        });
    }
}

let simba = new Animal('Simba', 'cat', 'Siamese');
let oscar = new Animal('Oscar', 'cat', 'Maine Coon');
let oliver = new Animal('Oliver', 'cat', 'British Shorthair');
let elvis = new Animal('Elvis', 'dog', 'Siberian Husky');
let buddy = new Animal('Buddy', 'dog', 'Retriever');
let bruno = new Animal('Bruno', 'dog', 'Alabai');

let mainGang = new AnimalGang('Animals');
let catGang = new AnimalGang('Cats');
let dogGang = new AnimalGang('Dogs');

mainGang.add(simba);
mainGang.add(elvis);
catGang.add(oscar);
catGang.add(oliver);
mainGang.add(catGang);
dogGang.add(buddy);
dogGang.add(bruno);
mainGang.add(dogGang);

mainGang.getDescription();
